ImageStats : Optimize by caching tiles and parallelizing

danieldresser-ie opened the pull request:
This is a fairly straightforward way to address: https://github.com/GafferHQ/gaffer/issues/3787

Main thing I haven't worked out yet is what type to use for the cache plugs.  I had thought to just use V3fPlug, but I hadn't realized there's no atomic version of it.  I quickly switched to using AtomicBox3f just to confirm that everything was working, but this is obviously ridiculous, and we probably should continue using double for the sums ( I did manage to construct a case with a tiny amount of visible error from using float, though it was a ridiculous case ).

I feel like I've run into this before, but do we have any good way of storing an arbitrary blob for internal data within a node ( without creating custom plug types for every case )?  It always really annoys me to use CompoundData for something that should be a struct - a lot more complexity and code, just to make things slower.  I guess we could just use IntVectorData to store blobs, and reinterpret_cast as a struct.  Would you find that objectionable?

Performance wise, this seems to help, though not quite as much as I was hoping.  Running all image tests went from 123 to 116.  But hopefully, any really expensive image tests won't be forced to single thread by assertImagesEqual now ( haven't tried combining this with the deep resampling PR yet ).  Interactive performance was definitely better - though still not as much as I would have hoped.  I might try profiling this.

johnhaddon commented on src/GafferImage/ImageStats.cpp:
```diff
@@ -198,98 +222,219 @@ const DeepState *ImageStats::deepState() const
+	if( output == tileStatsPlug() )
+	{
+		Imath::V2i tileOrigin = context->get<Imath::V2i>( ImagePlug::tileOriginContextName );
+		const Imath::Box2i tileBound = BufferAlgo::intersection(
+			Imath::Box2i( boundsIntersection.min - tileOrigin, boundsIntersection.max - tileOrigin ),
+			Imath::Box2i( Imath::V2i( 0 ), Imath::V2i( ImagePlug::tileSize() ) )
+		);
```
> Implement the final computation using parallelGatherTiles() or parallelProcessTiles(), pulling on the internal tile-stats plug for any tiles completely overlapped by the sampling region. (from original issue description)

I don't think I communicated this very well. I was thinking that `tileStatsPlug()` would contain the stats for the entire valid region of the tile, not the region clipped by `areaPlug()`. This would make interactive updates much quicker when changing the area, because we could reuse the stats for all tiles which are completely enclosed by the area. Only the final `allStatsPlug()` compute would depend on area, and that would be responsible for dealing with the clipped tiles at the edges.

johnhaddon replied:
I'd misread this - we're not including `areaPlug()` directly in the hash, but rather it's intersection with the tile bound. So we are indeed able to reuse cached results, as demonstrated in [this new test](https://github.com/GafferHQ/gaffer/pull/3792/commits/b490ad0c911f47d7e913154008db09cea6371bff#diff-441a2914869fe7e350ab2ba455418e1fR166).

johnhaddon commented on src/GafferImage/ImageStats.cpp:
```diff
@@ -198,98 +222,219 @@ const DeepState *ImageStats::deepState() const
+				sum += v.min[2];
+			},
+			boundsIntersection,
+			ImageAlgo::BottomToTop
```
Why BottomToTop?

johnhaddon replied:
From separate discussion : should have been `TopToBottom`, but we do need to avoid `Unordered`. For hash, we need a deterministic order to have a repeatable hash. For compute, floating point results would differ with different orders of addition.

johnhaddon commented on src/GafferImage/ImageStats.cpp:
```diff
@@ -198,98 +222,219 @@ const DeepState *ImageStats::deepState() const
+	else if(
+		input == allStatsPlug() ||
+		input == flattenedInPlug()->channelNamesPlug()
+	)
```
We still need `|| input == channelsPlug()` here, as the compute pulls on it indirectly via the `channelName()` method. Unless it hides significant complexity, I'm not a fan of "helpers" like `channelName()` when they hide plug accesses because they encourage bugs like this. I think `colorIndex()` and `channelName()` could be collapsed into a single static helper method which is passed plugs and/or values directly so it doesn't hide any plug accesses.

johnhaddon commented on src/GafferImage/ImageStats.cpp:
```diff
@@ -198,98 +222,219 @@ const DeepState *ImageStats::deepState() const
 	{
-		// Not a plug we know about
+		const int colorIndex = ::colorIndex( output );
+		if( colorIndex == -1 )
+		{
+			// Not a plug we know about
+			return;
+		}
```
I know this comes from the original code, but it's unnecessarily defensive. If anything it should be an assert in `colorIndex()` because there is no way we should ever hit this. If we do the refactor of `colorIndex/channelName` I suggested above, let's clean this up too.

danieldresser-ie commented:
OK, I think I've addressed all comments.  The main question is performance.

I did add the TaskCollaboration policy to allTasks, and saw a sizeable improvement when driving a Constant with the output color.  That much seems straightforward.

In terms of the benefits we would hope to see, I can very substantial improvements of the sort I had hoped to see ... as long as there's no cross-talk in the upstream hashing.  If the input is directly from a file read or an expensive OSLObject ( with a slow compute but a fast hash ), then interactively sliding the area is quite snappy ( the way we would hope it would be by caching tiles where the bound didn't change - I also added a test that the hash is performing correctly ).

Where the performance gets terrible is if there's a large blur or resize upstream.  Then the compute time gets completely dominated by hashing.  Trying to profile doesn't really show anything sticking out, it just a huge mess of little hash functions.  I guess it makes sense that nothing improves dramatically in these cases - the intermediate cache doesn't save any hashing time.  I would maybe have expected a bit more change from parallelizing the hashing - but perhaps it ends up repeatedly computing the same hashes?

To me, it feels like this is probably a reasonable step forward for now.  Seems like it might not be possible to get interactive dragging of the area feeling smooth in the general case without some sort of solution to the "whole-image-hash" thing we've talked about - stats are a textbook example of depending on the whole image ... though wait a minute, why isn't the hash cache making this better?

johnhaddon commented on src/GafferImage/ImageStats.cpp:
```diff
@@ -198,116 +236,222 @@ const DeepState *ImageStats::deepState() const
+			// Tile
+			[this] ( const ImagePlug *imageP, const Imath::V2i &tileOrigin )
+			{
+				return boost::static_pointer_cast<const IECore::V3dData>( tileStatsPlug()->getValue() )->readable();
```
I'm not entirely sure how this interacts with the result passing in the TBB pipeline, but this is returning a reference to data whose ownership is dropped as soon as we return. If the TBB pipeline doesn't take a copy (and I don't _think_ it will), we can get crashes if the value is removed from the cache before the gather completes. To avoid doubt, let's either explicitly pass V3d or V3dDataPtr by value.

johnhaddon replied:
I guess the other factor is the rules for automatic deduction of lambda return types, but since I don't know those off the top of my head either, I think something explicit would be preferable.

johnhaddon commented:
One other thing - could you update Changes.md with something like "ImageStats : Parallelised compute for improved performance"?

danieldresser-ie commented:
OK, I think that's feedback addressed.  I did the hashComputePolicy, but after spending a while trying was unable to come up with a test that showed any improvement.

I think I may have found a test showing 2-5% improvement, but the noise in this test was also in the 2-5% range, making it hard to reach much of a conclusion.  If this test was showing an improvement, and I'm not imagining things, then using TaskIsolation seems to show a similar benefit.  So something isn't working out the way we would hope, but I think it would take some pretty deep investigation to understand the hashing performance stuff.  Is it reasonable to leave it set to TaskCollaboration for the moment?  I didn't see any negative results, just a lot of ambiguity.

johnhaddon commented:
> Is it reasonable to leave it set to TaskCollaboration for the moment? I didn't see any negative results, just a lot of ambiguity.

Yep, I think so. I have some thoughts on how we might be able to rejig things to take better advantage of it.
